### PR TITLE
64-epoch-boundary cleanups and tests

### PR DESCRIPTION
**Pull request by djrtwo**
Tests green.

This could probably use some fork specific tests that test more explicitly period boundary and non-period boundary rewards/penalties

TODO: update test vector format files

### CODE REVIEW

File: specs/lightclient/beacon-chain.md [hunk 1]
```diff
@@ -630,28 +689,10 @@ def process_penalties(state: BeaconState) -> None:
                 state.leak_score[index] -= leak_score_decrease
 
     # Penalty processing
-    if get_current_epoch(state) % EPOCHS_PER_ACTIVATION_EXIT_PERIOD != 0:
```
Comment (terencechain) [inline specs/lightclient/beacon-chain.md:633]: By removing this, we will reset `leak_epoch_counter` to 0 every epoch. Is that intended?

**Review by hwwhww — COMMENTED** [review]
Great work!

---

We may need some more tests for the inactivity leak:

- The phase0 `@leaking` tests were generated by`transition_state_to_leak` function.
https://github.com/ethereum/eth2.0-specs/blob/6f2c69e79fd31815f8244d619f3487aff48004ad/tests/core/pyspec/eth2spec/test/helpers/rewards.py#L220-L226
    - It transitions to the epoch where `is_in_inactivity_leak(state) is True`. However, for HF1, the rewards and penalties are applied only on the period boundaries so the balance difference may not show on the test vectors.
- We can either (i) update `transition_state_to_leak` to something like your `transition_state_and_finalize_to_process_queue`, or (ii) create a new `@leaking_hf1` decorator for the HF1 specific tests.

File: specs/lightclient/beacon-chain.md (hunk 1 above)
Comment (djrtwo) [reply specs/lightclient/beacon-chain.md:633]: ah, right. I'll fix

Comment (djrtwo) [reply specs/lightclient/beacon-chain.md:633]: fixed and added test https://github.com/ethereum/eth2.0-specs/pull/2194/commits/56d3a30a054c6ca44dc3d7e5d8ff33378d064897

File: specs/lightclient/beacon-chain.md [hunk 2]
```diff
@@ -334,6 +336,61 @@ def get_flag_rewards(state: BeaconState, flag: ValidatorFlag, numerator: uint64)
...
+    """
+    penalties = [Gwei(0)] * len(state.validators)
+
+    # Only give penalties on period boundaries
+    if get_current_epoch(state) % EPOCHS_PER_ACTIVATION_EXIT_PERIOD != 0:
+        return penalties
+
+    for index in get_eligible_validator_indices(state):
+        # Inactivity-leak-specific penalty
+        if state.leak_score[index] >= EPOCHS_PER_ACTIVATION_EXIT_PERIOD * LEAK_SCORE_BIAS:
```
[the remark this answers is not in the document]
Comment (djrtwo) [reply specs/lightclient/beacon-chain.md:387]: It should definitely be 64. lol

chceking on the other comment

[the remark this answers is not in the document]
Comment (djrtwo) [reply specs/lightclient/beacon-chain.md:387]: The leak score for each validator is only set once per period in `process_penalties` https://github.com/ethereum/eth2.0-specs/pull/2194/files#diff-0e2661bd146a3f91f47225e75d80a3234e5ef4a0b451c57d6c34a822b2ee2f86R707 and just adds the bias times the number of leak epochs. (I don't believe it is the sum as you wrote it).

So, practically, you need to have more than an activation period's worth of leaks accumulated to get a leak penalty

File: tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_rewards_and_penalties.py [hunk 3]
```diff
@@ -27,6 +27,45 @@ def run_process_rewards_and_penalties(spec, state):
...
+                    assert post_state.balances[index] > pre_state.balances[index]
+                else:
+                    # Can only receive rewards on non-boundary epoch so neutral if not participanting
+                    assert post_state.balances[index] == pre_state.balances[index]
+            elif spec.EPOCHS_PER_ACTIVATION_EXIT_PERIOD > 2:
+                # Penalties always outweigh rewards on a period boundary if period > 2
+                assert post_state.balances[index] < pre_state.balances[index]
+            elif spec.EPOCHS_PER_ACTIVATION_EXIT_PERIOD == 2:
+                # Penalties either equal or outweigh rewards on period boundary == 2
+                assert post_state.balances[index] <= pre_state.balances[index]
```
[the remark this answers is not in the document]
Comment (djrtwo) [reply tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_rewards_and_penalties.py:66]: process_rewards can give maximally 2 base_rewards per epoch (a base reward for doing your job and a base reward that is canceled out in penalties). 

If the activation period is 2 epochs, then you'll have two penalties given during `process_penalties` to cancel out the reward penalties given during the period. So on the epoch transition epoch itself, you can get 2 == 2.

Note, if the activation period is > 2 epochs, the maximum reward in a single epoch is still 2 base_rewards but the cancellation penalties are for the entire period so you get `cancellation_penalties > rewards` for that single period transition epoch

**Comment by djrtwo** [comment]
@hwwhww I went with suggestion (i) for now. I would suspect we want to add more rewards/penalties tests for hf1 in another PR

File: tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_rewards_and_penalties.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (djrtwo) [reply tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_rewards_and_penalties.py]: notes added in comments
https://github.com/ethereum/eth2.0-specs/pull/2194/commits/dfce74e8797331309649120b6c6dd4be1469af8f#diff-bc4dda10c55f62aa97e54ebe53c38577075ff8145db4a9ea1727f50d8d60bc73R62-R65

**Pull request merged**